server/src/main/java/org/elasticsearch/tasks/TaskManager.java

```diff
@@ -607,4 +602,41 @@ void unregisterChildNode(DiscoveryNode node) {
         }
     }
 
+    /**
+     * Start tracking a cancellable task with its tcp channel, so if the channel gets closed we can get a set of
+     * pending tasks associated that channel and cancel them as these results won't be retrieved by the parent task.
+     *
+     * @return a releasable that should be called when this pending task is completed
+     */
+    public Releasable startTrackingCancellableChannelTask(TcpChannel channel, CancellableTask task) {
+        assert cancellableTasks.containsKey(task.getId()) : "task [" + task.getId() + "] is not registered yet";
+        final ChannelPendingTaskTracker tracker = channelPendingTaskTrackers.computeIfAbsent(channel, k -> new ChannelPendingTaskTracker());
```

Comment:
I think we need to clean these up when the channel closes. Maybe slightly tricky, we might close the channel concurrently with calling this method too.

Reply:
Good catch. I pushed https://github.com/elastic/elasticsearch/pull/56620/commits/cbe02c7eae5f3aa58f1a6757ac7697da7e95242f.